Fixing dockerd-rootless-setuptools.sh when user name contains a backslash

jmrouet opened the pull request:
active directory names like DOMAIN\user 
causes parsing issues because grep interprets the backslash

- Fixes https://github.com/moby/moby/issues/44396

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

jmrouet reviewed (commented):
Signed-off-by: Jean-Michel Rouet <jm.rouet@gmail.com>

thaJeztah commented:
Thank you for contributing! It appears your commit message is missing a DCO sign-off,
causing the DCO check to fail.

We require all commit messages to have a `Signed-off-by` line with your name
and e-mail (see ["Sign your work"](https://github.com/moby/moby/blob/v20.10.11/CONTRIBUTING.md#sign-your-work)
in the CONTRIBUTING.md in this repository), which looks something like:

```
Signed-off-by: YourFirsName YourLastName <yourname@example.org>
```

There is no need to open a new pull request, but to fix this (and make CI pass),
you need to _amend_ the commit(s) in this pull request, and "force push" the amended
commit.

Unfortunately, it's not possible to do so through GitHub's web UI, so this needs
to be done through the git commandline.

You can find some instructions in the output of the DCO check (which can be found
in the "checks" tab on this pull request), as well as in the [Moby contributing guide](https://github.com/moby/moby/blob/v20.10.11/docs/contributing/set-up-git.md).

Steps to do so "roughly" come down to:

1. Set your name and e-mail in git's configuration:

    ```bash
    git config --global user.name "YourFirstName YourLastName"
    git config --global user.email "yourname@example.org"
    ```

    (Make sure to use your _real_ name (**not your GitHub username/handle**) and e-mail)


2. Clone your fork locally
3. Check out the branch associated with this pull request
4. Sign-off and amend the existing commit(s)

    ```bash
    git commit --amend --no-edit --signoff
    ```

    If your pull request contains multiple commits, either squash the commits (if
    needed) or sign-off each individual commit.

5. _Force push_ your branch to GitHub (using the `--force` or [`--force-with-lease`](https://stackoverflow.com/questions/52823692/git-push-force-with-lease-vs-force) flags) to update the pull request.


Sorry for the hassle (I wish GitHub would make this a bit easier to do), and let me know if you need help or more detailed instructions!

AkihiroSuda commented on contrib/dockerd-rootless-setuptool.sh:
```diff
@@ -338,9 +341,12 @@ install_systemd() {
 	)
+	# fix usernames containing backslash like DOMAIN\user
+	# that will clash with display below
+	username=$(id -un | sed 's/\\/\\\\/g')
```
Can we make this a function to avoid code clone?

jmrouet replied:
Good idea to avoid duplicate code. I will add a USERNAME global variable.

jmrouet commented:
I signed-off the commits (thanks for the detailed explanations) and factorized the code as suggested.
Let me know if this can be merged then.

jmrouet committed:
more robust dockerd-rootless-setuptools.sh

Fixing case where username may contain a backslash.
This case can happen for winbind/samba active directory domain users.

Signed-off-by: Jean-Michel Rouet <jean-michel.rouet@philips.com>

Use more meaningful variable name

Signed-off-by: Jean-Michel Rouet <jean-michel.rouet@philips.com>

Update contrib/dockerd-rootless-setuptool.sh

Co-authored-by: Akihiro Suda <suda.kyoto@gmail.com>
Signed-off-by: Jean-Michel Rouet <jean-michel.rouet@philips.com>

Use more meaningful variable name

Signed-off-by: Jean-Michel Rouet <jean-michel.rouet@philips.com>

Update contrib/dockerd-rootless-setuptool.sh

Co-authored-by: Akihiro Suda <suda.kyoto@gmail.com>
Signed-off-by: Jean-Michel Rouet <jean-michel.rouet@philips.com>

jmrouet commented:
I cleaned up the messy commits on my fork, and signed-off correctly the single commit.
The cleanup may have closed the PR, so I'm trying to reopen.
Sorry for the confusion

AkihiroSuda commented:
Please squash commits

jmrouet commented:
> 
Commits squashed

thaJeztah commented:
Let me quote the comment from the ticket here as well; https://github.com/moby/moby/issues/44396#issuecomment-1300993995

> Having `\` in a username on a Linux system sounds like the kind of thing that will probably break a lot of tools 😬
> 
> It's bad enough that `adduser` even rejects it if you try to create one:
> 
> ```
> root@c06a555fcf1e:/# adduser 'foo\bar'
> adduser: To avoid problems, the username should consist only of
> letters, digits, underscores, periods, at signs and dashes, and not start with
> a dash (as defined by IEEE Std 1003.1-2001). For compatibility with Samba
> machine accounts $ is also supported at the end of the username
> ```
> 
> IMO we should only support this if the changes can be kept extremely minimal (since I don't think any of the maintainers here will be testing this use case to make sure it doesn't break). 🙈


@tianon ptal; let me know if you think this patch is "minimal" enough 😅 (I know it's a bit of an odd one to have such usernames on Linux)

jmrouet commented:
I see 2 failing checks in the CI/CD.
Is it independant from this PR ? 

I agree with @tianon's comment that having `\` in a username on a Linux system sounds weird, but this is how samba/windows let ActiveDirectory domain users log onto a unix server. 
In this configuration when they ssh into our server their login name appears as `DOMAIN\User`. Thereafter they cannot run `dockerd-rootless-setuptools.sh` unless they tweak the script as proposed in this PR.

To make things even more worse, in our Active Directory domain, IT admins chose to assign windows usernames made of digits only... something that is also really not recommended in the Unix world as it creates confusion between UID and username for commands that take both forms (like chown)...

AkihiroSuda commented on contrib/dockerd-rootless-setuptool.sh:
```diff
@@ -78,6 +79,10 @@ init() {
 
+	# Set USERNAME from `id -un` and potentially protect backslash
+	# for windbind/samba domain users
+	USERNAME=$(id -un | sed 's/\\/\\\\/g')
```
I feel this variable should be named like `USERNAME_ESCAPED`

jmrouet replied:
Good idea.
Just implemented it.

AkihiroSuda commented:
Please squash commits, then LGTM

AkihiroSuda commented on contrib/dockerd-rootless-setuptool.sh:
```diff
@@ -340,7 +347,7 @@ install_systemd() {
 	INFO "Installed ${SYSTEMD_UNIT} successfully."
 	INFO "To control ${SYSTEMD_UNIT}, run: \`systemctl --user (start|stop|restart) ${SYSTEMD_UNIT}\`"
-	INFO "To run ${SYSTEMD_UNIT} on system startup, run: \`sudo loginctl enable-linger $(id -un)\`"
+	INFO "To run ${SYSTEMD_UNIT} on system startup, run: \`sudo loginctl enable-linger ${USERNAME_ESCAPED}\`"
```
```suggestion
	INFO "To run ${SYSTEMD_UNIT} on system startup, run: \`sudo loginctl enable-linger ${USERNAME}\`"
```

AkihiroSuda commented:
ping @jmrouet

jmrouet commented:
Hello @AkihiroSuda 
I accepted your change. I didn't noticed this message was pending for so long. Sorry for that.
After accepting the commit from the gitlab interface, and locally signed-off and forced pushed the commit again. Is that ok?

AkihiroSuda commented:
Please squash commits into a single commit

jmrouet commented:
Ok so I did
$ git rebase -i HEAD~3 --signoff
kept the first line as "pick" and the other as "squash"
then
$ git puss --force-with-lease

Is that OK?

AkihiroSuda reviewed (approved):
Thanks
